### ISSUE TITLE
Line/Area/Bar charts incorrectly displayed as a timeseries with certain data

### ISSUE THREAD

**Issue by agilliland** [metabase/metabase#855]
We are missing a concrete data set to reproduce this, but I did see it for sure.  A user created a query that was a count grouped by a column which was alphanumeric (ex: `scanner005`) and when you tried to chart the results using a bar chart the x-axis was indicating time series, which was clearly wrong.  no actual bars showed up on the graph.

**Comment by tlrobinson** [comment]
@agilliland I'm having trouble reproducing this on MySQL or Postgres, even with alphanumeric column names. Is it possible the column was accidentally marked as a timestamp or something?

**Comment by agilliland** [comment]
I would just go with the top 2-3 standards, with ISO 8601 being the most important.

**Comment by camsaul** [comment]
Why don't we just look at the column type info returned with the results and only try to parse it if it's a `:DateField` / `:DateTimeField` / etc.?

**Comment by tlrobinson** [comment]
Good question. [That code](https://github.com/metabase/metabase/blob/master/frontend/src/card/card.charting.js#L77-L83) is super old, so maybe it's no longer necessary?

**Comment by agilliland** [comment]
I think it's mainly a question of whether or not we will always have a base-type that we can rely on.  So I believe the main reason we tried the parsing was for SQL queries, which can output their results such that the values are strings even though they represent dates.

**Comment by camsaul** [comment]
@tlrobinson it looks like the code above it isn't checking for a few other types like `:DateTimeField`. At any rate the QP always returns dates in ISO 8601/RFC 3339 and every SQL database I can think of does as well so I would only try to parse that format and nothing else.

**Comment by tlrobinson** [comment]
Ok, I propose something like `isDate(result.cols[0]) || moment(results.rows[0][0], [moment.ISO_8601, moment.RFC_3339]).isValid()`;

`isDate` is defined here: https://github.com/metabase/metabase/blob/master/frontend/src/lib/schema_metadata.js#L22-L25 @camsaul is it safe to assume the QP will return a ISO 8601 date for those base/special types?

`moment.ISO_8601` is included with moment.js, `moment.RFC_3339` is not but we can define it.

If we want to be more robust we could check every value, rather than just first.

**Comment by camsaul** [comment]
Yup, those types are good. I wouldn't sweat adding a separate RFC 3339 parser. TL;DR it's basically a slightly stricter subset of ISO 8601, so the ISO 8601 parser should handle everything that comes back.

**Comment by agilliland** [comment]
I don't think we need to try this on every value.  If it fails on the first one then I think it's fine to assume it's not a DT value.  Most of the time the metadata will be our guide, the value check is just a very simple fallback.

**Comment by camsaul** [comment]
But also keep in mind the first value for a column could be `nil` so a SQL query in this case wouldn't get the correct type info. It's a PITA but you should really scan until you find a non-nil value

**Comment by camsaul** [comment]
Wait a minute. I wrote code on the backend to do this already. Type information comes back with SQL queries 

https://github.com/metabase/metabase/blob/master/src/metabase/driver/generic_sql/native.clj#L43-L45

**Comment by agilliland** [comment]
For sure.  I was just going to skip that case because it's going to be a really minute set of cases where we couldn't determine the column type via metadata and it has valid date values but the first one is nil.

**Comment by camsaul** [comment]
The backend doesn't try test whether strings are datetimes or not. With most DBs they come back as `Dates` or `Timestamps` anyway which it translates to the correct type. SQLite is a notable exception if I remember correctly and always returns dates as strings. We can add code to test strings pretty easily.

Anyway... the backend code is where this code belongs in the first place; since it's already there go ahead and remove the duplicate logic from the frontend. If we need more aggressive type inference we can add it to the backend where we can test it

**Comment by agilliland** [comment]
I'm not sure that we should be quite so aggressive about manipulating the values in the backend.  Specifically when users are running native queries I don't think we should be transforming the values on the output of their queries because it's entirely likely that they formatted the values in a specific way on purpose.

I'm okay with being more liberal about this when it's a structured query and we are the ones writing the SQL and it's expected that we are managing the query, but we shouldn't mess with the output of people's native queries.

**Comment by camsaul** [comment]
Who said anything manipulating values? We're talking about how automatic type inference is done.

**Comment by agilliland** [comment]
When you said `We could add code to test strings pretty easily if we want it.` I interpreted that as you felt we should do in the QP what the frontend is currently doing and potentially parse string values to see if they are actually datetime values, and if so then we would format them accordingly.

If that's not what you meant then I misunderstood and just need some more clarification to know what you meant.

When you say we should remove the duplicate logic from the frontend i'm not clear what code you are talking about then.  The code on the frontend only does 2 things, first it attempts to see if the metadata indicates the data is a DT field and if so just accepts that, otherwise it will make a very simple attempt to parse the first value of data and see if it's a DT, which i think we should continue doing.  So i don't see any real duplication.

**Comment by camsaul** [comment]
When I said we can testing a string to see if it's a datetime I meant just that. We can test it and add the correct type information in the response so the frontend doesn't need to try to parse it too. That way the frontend just needs to check the tag which makes the logic a lot simpler there.

**Comment by joshuashort** [comment]
I'm using v0.24.1 with Docker, and I found this issue while searching for "metabase bar visualization wrong grouping"..

I'm aggregating on a column named `Part Num`, summing a field that gets rendered as `Q Ty` (i.e. Qty), and the table view is correct, but the bar graph is labeling the parts as dates.

If it helps, the first part numbers include '1002-08280', '1002-08300', etc., in a vaguely date-like format (but some are things like '5001-MI654' and 'SAN_SDJCR 1212K 11-S').

Am I missing something?

**Comment by tlrobinson** [comment]
Is this a SQL query or using the graphical query editor?

**Comment by joshuashort** [comment]
The visual editor -- I only looked at the generated query.

**Comment by joshuashort** [comment]
FWIW, the field type was originally "Category", and I also tried with it as "No field type".

On a related note, would a date/time field type fix this auto detection(?) issue via explicitness? Then a chart could know it's certainly a date.. without whatever guessing it's doing that is out of sync with the (correct) result in a table visualization?

**Comment by joshuashort** [comment]
I'm still getting this issue with v0.24.2 (and on every master build I've checked). I can reliably reproduce it..

If there are the following part numbers in a table: "075030", "2002-50060", "4001-00450" (and some corresponding quantity for each), the Bar Chart works when filtering on the fist two. Trying to filter by "Starts With: 4001-" causes the x-axis to turn into a date. I've tried setting the field type to Entity Name, Category, None, Description.. nothing helps.

Please, someone, what's going on here? Is it something with MariaDB? I noticed I can't run Maria as the storage for Metabase with the mysql driver..

**Comment by joshuashort** [comment]
Sweeeeet! I found out where to short-circuit (and undoubtedly break something I haven't run into yet) the timeseries test that is giving me faulty visualizations -- the selected function is at https://github.com/metabase/metabase/blob/ad15dcee4878d57c03ff3a98a9a24a55358c4cb1/frontend/src/metabase/visualizations/lib/timeseries.js#L20-L25

I just replaced the return logic with `false` and now the visualizations show up correctly. An actual fix seems like it'd involve `/frontend/src/metabase/visualizations/lib/LineAreaBarRenderer.js` or `src/frontend/src/metabase/lib/schema_metadata.js`, but I don't have enough familiarity to know what's actually going on. Looking at the history here, I hope there's a better approach to this problem in the works..
